### ISSUE TITLE
False negatives and false positives for "Object is possibly 'null'" warning in typescript@next with "complex" if 

### ISSUE THREAD

**Issue by donaldpipowitch** [Microsoft/TypeScript#10198]
**TypeScript Version:**  2.1.0-dev.20160807

**Code**

```ts
class Nested {
  data: string;
}

class Foo {
  bar: Nested | null = null;
  baz: number = 1;

  get data(): null | string {
    if (this.bar === null || this.baz !== 1) {
      return null;
    } else {
      return this.bar.data;
             // ^~~~~ Object is possibly 'null'. (BUT IT CAN'T BE NULL.)
    }
  }
}
```

**Expected behavior:**

No error.

**Actual behavior:** 

An error.

**Details:** 

This failure must be new. I think I installed `typescript@next` last week and had no error. I installed it today and it throws an error now.

If a you remove the second if-branch (`this.baz !== 1`) it works.

This also throws:

```
  get data(): null | string {
    if (this.bar === null) {
      return null;
    } else if (this.baz !== 1) {
      return null;
    } else {
      return this.bar.data;
             // ^~~~~ Object is possibly 'null'. (BUT IT CAN'T BE NULL.)
    }
  }
```

This works:

```
  get data(): null | string {
    if (this.baz !== 1) {
      return null;
    } else if (this.bar === null) {
      return null;
    } else  {
      return this.bar.data;
    }
  }
```

(!) On the other hand: This works, but could be wrong so it **should** throw an error!

```
  get data(): null | string {
    if (this.baz !== 1 || this.bar === null) {
      return null;
    } else {
      return this.bar.data;
    }
  }
```

**Comment by yortus** [comment]
This also works:

```ts
  get data(): null | string {
    let x = this.bar; // x is Nested | null
    if (x === null || this.baz !== 1) {
      return null;
    } else {
      return x.data; // OK
    }
  }
```

**Comment by ahejlsberg** [comment]
Duplicate of #10169, fix available in #10188.
